Mathlib/Combinatorics/Graph/Basic.lean

```diff
@@ -0,0 +1,323 @@
+
+lemma IsLink.isLink_iff (h : G.IsLink e x y) {x' y' : α} :
+    G.IsLink e x' y' ↔ (x = x' ∧ y = y') ∨ (x = y' ∧ y = x') := by
+  refine ⟨h.eq_and_eq_or_eq_and_eq_of_isLink, ?_⟩
+  rintro (⟨rfl, rfl⟩ | ⟨rfl,rfl⟩)
+  · assumption
+  exact h.symm
+
+lemma IsLink.isLink_iff_sym2_eq (h : G.IsLink e x y) {x' y' : α} :
+    G.IsLink e x' y' ↔ s(x,y) = s(x',y') := by
+  rw [h.isLink_iff, Sym2.eq_iff]
+
+/-! ### Edge-vertex incidence -/
+
+/-- The unary incidence predicate of `G`. `G.Inc e x` means that the vertex `x`
+is one or both of the ends of the edge `e`. -/
+def Inc (G : Graph α β) (e : β) (x : α) : Prop := ∃ y, G.IsLink e x y
+
+@[simp]
+lemma Inc.edge_mem (h : G.Inc e x) : e ∈ E(G) :=
+  h.choose_spec.edge_mem
+
+@[simp]
+lemma Inc.vertex_mem (h : G.Inc e x) : x ∈ V(G) :=
+  h.choose_spec.left_mem
+
+@[simp]
+lemma IsLink.inc_left (h : G.IsLink e x y) : G.Inc e x :=
+  ⟨y, h⟩
+
+@[simp]
+lemma IsLink.inc_right (h : G.IsLink e x y) : G.Inc e y :=
+  ⟨x, h.symm⟩
+
+lemma Inc.eq_or_eq_of_isLink (h : G.Inc e x) (h' : G.IsLink e y z) : x = y ∨ x = z :=
+  h.choose_spec.left_eq_or_eq_of_isLink h'
+
+lemma Inc.eq_of_isLink_of_ne_left (h : G.Inc e x) (h' : G.IsLink e y z) (hxy : x ≠ y) : x = z :=
+  (h.eq_or_eq_of_isLink h').elim (False.elim ∘ hxy) id
+
+/-- The binary incidence predicate can be expressed in terms of the unary one. -/
+lemma isLink_iff_inc : G.IsLink e x y ↔ G.Inc e x ∧ G.Inc e y ∧ ∀ z, G.Inc e z → z = x ∨ z = y := by
+  refine ⟨fun h ↦ ⟨h.inc_left, h.inc_right, fun z h' ↦ h'.eq_or_eq_of_isLink h⟩, ?_⟩
+  rintro ⟨⟨x', hx'⟩, ⟨y', hy'⟩, h⟩
+  obtain rfl | rfl := h _ hx'.inc_right
+  · obtain rfl | rfl := hx'.left_eq_or_eq_of_isLink hy'
+    · assumption
+    exact hy'.symm
+  assumption
+
+/-- Given a proof that the edge `e` is incident with the vertex `x` in `G`,
+noncomputably find the other end of `e`. (If `e` is a loop, this is equal to `x` itself). -/
+noncomputable def Inc.other (h : G.Inc e x) : α := h.choose
+
+@[simp]
+lemma Inc.isLink_other (h : G.Inc e x) : G.IsLink e x h.other :=
+  h.choose_spec
+
+@[simp]
+lemma Inc.inc_other (h : G.Inc e x) : G.Inc e h.other :=
+  h.isLink_other.inc_right
+
+lemma Inc.eq_or_eq_or_eq_of_inc_of_inc (hx : G.Inc e x) (hy : G.Inc e y) (hz : G.Inc e z) :
+    x = y ∨ x = z ∨ y = z := by
+  by_contra! hcon
+  obtain ⟨x', hx'⟩ := hx
+  obtain rfl := hy.eq_of_isLink_of_ne_left hx' hcon.1.symm
+  obtain rfl := hz.eq_of_isLink_of_ne_left hx' hcon.2.1.symm
+  exact hcon.2.2 rfl
+
+/-- `G.IsLoopAt e x` means that both ends of the edge `e` are equal to the vertex `x`. -/
+def IsLoopAt (G : Graph α β) (e : β) (x : α) : Prop := G.IsLink e x x
+
+@[simp]
+lemma isLink_self_iff : G.IsLink e x x ↔ G.IsLoopAt e x := Iff.rfl
+
+lemma IsLoopAt.inc (h : G.IsLoopAt e x) : G.Inc e x :=
+  IsLink.inc_left h
+
+lemma IsLoopAt.eq_of_inc (h : G.IsLoopAt e x) (h' : G.Inc e y) : x = y := by
+  obtain rfl | rfl := h'.eq_or_eq_of_isLink h <;> rfl
+
+@[simp]
+lemma IsLoopAt.edge_mem (h : G.IsLoopAt e x) : e ∈ E(G) :=
```

Comment:
```suggestion
lemma IsLoopAt.mem_edgeSet (h : G.IsLoopAt e x) : e ∈ E(G) :=
```
is more correct and probably more discoverable too

Reply:
This one follows the convention established in `Inc.edge_mem`, `IsLink.edge_mem`, etc.

Reply:
I didn't comment on this one earlier because it's in parallel to the IsLink variant. Arguably `IsLoopAt` has the same naming considerations as what's documented about IsLink, so saying it's more correct isn't necessarily correct (though it's not documented that `IsLoopAt` uses the same naming scheme as `IsLink`).

I personally slightly prefer `mem_edgeSet` rather than use `edge`.